Annotato/Annotato/Views/Document/Edit/DocumentView.swift

```diff
@@ -16,16 +18,105 @@ class DocumentView: UIView {
+        NotificationCenter.default.addObserver(
+            self, selector: #selector(handleVisiblePageChange(notification:)),
+            name: Notification.Name.PDFViewVisiblePagesChanged, object: nil
+        )
+    }
+
+    @objc
+    private func handleVisiblePageChange(notification: Notification) {
+        initializeAnnotationViewsForVisiblePages()
+    }
+
+    private func checkIfAnnotationIsInVisiblePages(
+        anno: DocumentAnnotationViewModel,
+        visiblePages: [PDFPage]
+    ) -> Bool {
+        guard let pageNumForAnnotation = anno.associatedPage.label else {
+            return false
+        }
+        let visiblePagesIndex = visiblePages.map({ pdfPage -> String in
+            guard let label = pdfPage.label else {
+                return "-1"
+            }
+            return label
+        })
+        return visiblePagesIndex.contains(pageNumForAnnotation)
+    }
+
+    private func checkIfAnnoIsAlreadyInSubview(
+        anno: DocumentAnnotationViewModel, subviews: [UIView]
+    ) -> Bool {
+        for subview in subviews {
+            guard let annoSubview = subview as? DocumentAnnotationView else {
+                continue
+            }
+            if annoSubview.viewModel === anno {
+                return true
+            }
+        }
+        return false
+    }
+
+    private func bringAnnoToFront(
+        anno: DocumentAnnotationViewModel, subviews: [UIView]
+    ) {
+        for subview in subviews {
+            guard let annoSubview = subview as? DocumentAnnotationView else {
+                continue
+            }
+            if annoSubview.viewModel === anno {
+                guard let documentViewParent = annoSubview.superview else {
+                    return
+                }
+                documentViewParent.bringSubviewToFront(annoSubview)
+            }
+        }
+    }
+
+    private func initializeAnnotationViewsForVisiblePages() {
+        guard let pdfSubView = pdfView else {
+            return
+        }
+        let visiblePages = pdfSubView.visiblePages
         for annotation in documentViewModel.annotations {
             let view = DocumentAnnotationView(annotationViewModel: annotation)
-            addSubview(view)
+            let annoShouldBeVisible = checkIfAnnotationIsInVisiblePages(anno: annotation, visiblePages: visiblePages)
+            guard let documentView = pdfSubView.documentView else {
+                continue
+            }
+
+            if annoShouldBeVisible {
+                let annoIsAlreadyInSubview = checkIfAnnoIsAlreadyInSubview(
+                    anno: annotation, subviews: documentView.subviews
+                )
+                if !annoIsAlreadyInSubview {
+                    pdfSubView.documentView?.addSubview(view)
+                } else {
+                    bringAnnoToFront(
+                        anno: annotation, subviews: documentView.subviews
+                    )
+                }
+            }
```

Comment:
If this is purely a view thing, why can't we simply `bringAnnoToFront(annotationView)`?
1. It seems to mess up responsibility if we have to perform logic on view models (by the equality comparison) in the view layer.
2. It requires a lot more if-else logic which would be difficult to ensure correctness

But please correct me if I missed out your reason for doing it this way

Reply:
I don't really understand this, my bad

This function iterates all the subviews, and checks uses the equality comparison to get the correct subview (type=UIView) as? AnnotationView

Then we can find out whether it is in the visible pages
Then we can decide whether to bring it to front (front as in top of the zstack)

If I shift this somewhere else I'll still have to reach into the subviews
Because as I said above, the parent of the annotation view is the innermost documentView that is defined in PDFKit
We cannot include any logic directly there if I'm not wrong